[Docs] Add example of custom actions use with custom components (#392)
mckinsey/vizro: vizro-core/docs/pages/user-guides/custom-components.md

    [CustomComponent2]: ../../assets/user_guides/custom_components/customcomponent_2.png

# Custom component with custom action

Custom components can be used as inputs to custom actions.

> Following the above instructions to create a custom component, we ended with the below `OffCanvas` component.

Reviewer: ```suggestion
Following the instructions above to create a custom component, results in this `OffCanvas` component:
```

```py
class OffCanvas(vm.VizroBaseModel):
    type: Literal["offcanvas"] = "offcanvas"
    title: str
    content: str

    def build(self):
        return html.Div(
            [
                dbc.Offcanvas(
                    children=html.P(self.content),
                    id=self.id,
                    title=self.title,
                    is_open=False,
                ),
            ]
        )
```

> To enable our custom component to work with our custom action, we need to add few additional lines of code.

Reviewer: ```suggestion
To enable the custom component to work with the custom action, we need to add some additional lines of code.
```

> 1. First, you have to add `actions` argument to your custom component. Type of `actions` argument is `List[Action]`.

Reviewer: ```suggestion
1. **Add the `actions` argument to your custom component**. The type of the `actions` argument is `List[Action]`.
```

> 2. Next step is to define `_input_property` of the component. `_input_property` is the property that is used as the input to the custom action callback

Reviewer: ```suggestion
2. **Define the `_input_property` of the component**. The `_input_property` is the property used as the input to the custom action callback
```

   ```py
    _input_property: str = PrivateAttr("is_open")
   ```

> 3. Next, you need to `_set_actions`. By setting action, change in 'is_open' property of our custom component will trigger the action.

Reviewer: ```suggestion
3. **Set the action through `_set_actions`**. In doing so, any change in the `is_open` property of the custom component triggers the action.
```

   ```py
    _set_actions = _action_validator_factory("is_open")

>    ```

Reviewer 2: None of these are required to enable a custom component (property) to be the `input`/`output` of a custom action.

Some of these requirements (`1.` and `3.`) are really needed if the custom component `triggers` a custom action. (`2.` isn't required even then, because `_input_property` only stands for component's `property` which will be used if the component is used as the filter/parameter selector).

However, I'd really like to see another example below that describes how to configure a custom component to `trigger` an action (custom or predefined). Then you can add requirements (`1.` and `3.`) and an example. 
In that case, the title `"## Custom component with custom action"` could be renamed to something like `"## Custom component with actions"` which could contain these two subtitles:
 - "### Custom component as input/output of custom action"
 - "### Triggering actions with a custom component"

Reviewer: Sorry, I didn't see this comment above before I made the edits. I've just edited for style to bold the first sentence of each instruction and make it imperative, so it's clearer for people to swiftly scan the list. If you make changes for Petar's comments, please just follow that guideline.

Author: Removed this section for offcanvas component. Added another example for when the custom component triggers the action.

> After you have completed the above steps, it is time to write your custom [action]('user-guides/custom-actions.md').

Reviewer: ```suggestion
After you have completed the steps above, it is time to write your custom [action]('user-guides/custom-actions.md').
```

   ```py
    @capture("action")
    def open_offcanvas(n_clicks, is_open):
        if n_clicks:
            return not is_open
        return is_open
   ```

> Add our custom action `open_offcanvas` as a `function` argument inside the [`Action`][vizro.models.Action] model.

Reviewer: ```suggestion
Add the custom action `open_offcanvas` as a `function` argument inside the [`Action`][vizro.models.Action] model.
```

        vm.Page.add_type("components", OffCanvas)

        # 3. Create custom action
        @capture("action")
        def open_offcanvas(n_clicks, is_open):
            if n_clicks:
                return not is_open
            return is_open

        page = vm.Page(
            title="Custom Component",
            components=[
                vm.Button(
                    text="Open Offcanvas",
                    id="open_button",
                    actions=[
                        vm.Action(
                            function=open_offcanvas(),
                            inputs=["open_button.n_clicks", "offcanvas.is_open"],
                            outputs=["offcanvas.is_open"],
                        )
                    ],
                ),
                OffCanvas(
                    id="offcanvas",
                    content="OffCanvas content",
                    title="Offcanvas Title",
                ),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])

        Vizro().build(dashboard).run()

        ```

        1.  Here we set the property that is used as the input to the custom action callback.

>         2.  Here we set action. Change in 'is_open' property of our custom component will trigger the action.

Reviewer: ```suggestion
        2.  Here we set the action so a change in the `is_open` property of the custom component triggers the action.
```

    [CustomComponent2]: ../../assets/user_guides/custom_components/customcomponent_2.png

## Custom component with custom action

> Custom components can be used as inputs to, or outputs of custom actions. In the examples below we will explore both options.

Reviewer 2: ```suggestion
Custom components can be used as `inputs` to, `outputs` of, or as a `trigger` of custom actions. In the examples below we will explore both options.
```

            return is_open

        page = vm.Page(
            title="Custom Component",
            components=[
                vm.Button(
                    text="Open Offcanvas",
                    id="open_button",
                    actions=[
                        vm.Action(
                            function=open_offcanvas(),
                            inputs=["open_button.n_clicks", "offcanvas.is_open"],
                            outputs=["offcanvas.is_open"],
                        )
                    ],
                ),
                OffCanvas(
                    id="offcanvas",
                    content="OffCanvas content",
                    title="Offcanvas Title",
                ),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])

        Vizro().build(dashboard).run()

        ```
    === "yaml"
        ```yaml
        # Custom components are currently only possible via python configuration
        ```
    === "Result"
        [![CustomComponent3]][CustomComponent3]

    [CustomComponent3]: ../../assets/user_guides/custom_components/customcomponent_3.gif

### Triggering actions with a custom component

> As mentioned above, custom components can trigger action. To enable the custom component to trigger the action, we need to add some additional lines of code:
> 
> To enable the custom component to work with the custom action, we need to add some additional lines of code:

Reviewer 2: ```suggestion
As mentioned above, custom components can also trigger action. To enable the custom component to trigger the action, we need to add some additional lines of code:
```

> 1. **Add the `actions` argument to your custom component**. The type of the `actions` argument is `List[Action]`.
> 2. **Set the action through `_set_actions`**. In doing so, any change in the `active_index` property of the custom component triggers the action.

Reviewer 2: ```suggestion
1. **Add the `actions` argument to your custom component**. The type of the `actions` argument is `List[Action]`.
   ```py
   actions: List[Action] = []
   ```
2. **Set the action through `_set_actions`**. In doing so, any change in the `"active_index"` property of the custom component triggers the action.
```

   ```py
    _set_actions = _action_validator_factory("active_index")
   ```

??? example "Example of triggering action with custom component"

    === "app.py"
        ``` py
        from typing import List, Literal

        import dash_bootstrap_components as dbc
        import vizro.models as vm
        from dash import html
        from vizro import Vizro

        try:
            from pydantic.v1 import Field, PrivateAttr
        except ImportError:
            from pydantic import PrivateAttr

        from vizro.models import Action
        from vizro.models._action._actions_chain import _action_validator_factory
        from vizro.models.types import capture

        # 1. Create new custom component
        class Carussel(vm.VizroBaseModel):
        type: Literal["carussel"] = "carussel"
        items: List
        actions: List[Action] = []

        _set_actions = _action_validator_factory("active_index")  # (1)!

        def build(self):
            return dbc.Carousel(
                id=self.id,
                items=self.items,

>             )

Reviewer 2: Can you fix an indentation?

    [CustomComponent2]: ../../assets/user_guides/custom_components/customcomponent_2.png

> ## Custom component with custom action

Reviewer: Sorry, one more change to make the headings consistent

> ## How to use custom components with custom actions

Reviewer 3: Could we remove the redundants "how to" across this page to make the menu on the right easier to read?

Author: "How to" is used throughout the docs, so removing it here would make custom-components inconsistent with the rest of the docs. @stichbury what do you think? Should we remove the "how to" throughout the docs"?

Reviewer 3: You are correct, it is indeed used in other places - we are actually not really consistent I think as it varies. I may have forgotten what out current stand on this is. @stichbury could you remind us what we decided? I personally much prefer not having lots of redundant "How to" on the side, but I remember that we liked it as section headers...

Reviewer: This was my bad, sorry! I added the "How to" back in. I haven't codified the approach but I'll do this somewhere so I remember and can point other authors to it as needed.

So we have a "How-to guides" section and within that, a set of pages (see left hand ToC)

<img width="218" alt="Screenshot 2024-04-03 at 11 51 36" src="https://github.com/mckinsey/vizro/assets/5180475/bb80668e-1e96-4fb8-b7e5-a80476d17e74">

* Guidance 1: Each page should have a "How-to title" such as this one:

<img width="1038" alt="image" src="https://github.com/mckinsey/vizro/assets/5180475/9f8f12cf-bd0b-47b4-a5d0-67115b9a82e4">

* Guidance 2: Subsections on the page should not be "How to..." do different things. The subsections should instead be worded as imperative, so rather than "How to customize pre-defined actions" it should "Customize pre-defined actions". It should also not be gerund form (so don't use "Customizing pre-defined actions".

Purpose of these two "guidances" (note, not rules) is to keep the page clear about what it does (it's a how-to guide) but not overload the right hand ToC with loads of redundant and repetitive text:

<img width="262" alt="image" src="https://github.com/mckinsey/vizro/assets/5180475/52c014f2-e34e-41ee-9eaa-2369e4714c24">

@nadijagraca if you are OK to modify this file, leave me to sort out any other pages where it's currently not following guidance in a separate PR?

Custom components can be used as `inputs` to, `outputs` of, or as a `trigger` of custom actions. In the examples below we will explore both options.

> ### Custom component as input/output of custom action

Reviewer: ```suggestion
### Custom components as inputs/outputs of custom actions
```

Reviewer: This could also be a "how to" header but I suspect it'll make it long in the table of contents, so I'm happy to leave it like this (but use the same form as in the paragraph above, so changed to plurals here).

            if n_clicks:
                return not is_open
            return is_open

        page = vm.Page(
            title="Custom Component",
            components=[
                vm.Button(
                    text="Open Offcanvas",
                    id="open_button",
                    actions=[
                        vm.Action(
                            function=open_offcanvas(),
                            inputs=["open_button.n_clicks", "offcanvas.is_open"],
                            outputs=["offcanvas.is_open"],
                        )
                    ],
                ),
                OffCanvas(
                    id="offcanvas",
                    content="OffCanvas content",
                    title="Offcanvas Title",
                ),
            ],
        )

        dashboard = vm.Dashboard(pages=[page])

        Vizro().build(dashboard).run()

        ```
    === "yaml"
        ```yaml
        # Custom components are currently only possible via python configuration
        ```
    === "Result"
        [![CustomComponent3]][CustomComponent3]

    [CustomComponent3]: ../../assets/user_guides/custom_components/customcomponent_3.gif

> ### Triggering actions with a custom component

Reviewer: ```suggestion
### How to trigger actions with a custom component
```

As mentioned above, custom components can trigger action. To enable the custom component to trigger the action, we need to add some additional lines of code:

> 1. **Add the `actions` argument to your custom component**. The type of the `actions` argument is `List[Action]`.

Reviewer 3: Do we need bolding here?

Reviewer: I think so, it helps make it clear what the action (pardon the pun) in the sequence is.
